File: week_3/5_maximum_number_of_prizes/different_summands.cpp
#include <iostream>
#include <vector>

using std::vector;

vector<int> optimal_summands(int n) {
  vector<int> summands;

	uint64_t i = 1;
	while(n > 2*i){
		summands.emplace_back(i);
		n = n - i;
		++i;	
	}
	summands.emplace_back(n);
	//write your code here
  return summands;
}

int main() {
  int n;
  std::cin >> n;
  vector<int> summands = optimal_summands(n);
  std::cout << summands.size() << '\n';
  for (size_t i = 0; i < summands.size(); ++i) {
    std::cout << summands[i] << ' ';
  }
}

File: week_3/5_maximum_number_of_prizes/different_summands.hs
-- by Kirill Elagin

get_values :: Int -> Int-> [Int]
get_values n i 
        | n > 2*i = i:get_values (n-i) (i+1)
        | otherwise = return n



optimal_summands :: Int -> [Int]
optimal_summands n = get_values n 1 -- write your code here

main :: IO ()
main = do
  n  <- getLine
  let summands = optimal_summands (read n)
  print $ length summands
  putStrLn $ unwords (map show summands)
